dm/pkg/checker/conn_checker.go

```diff
@@ -0,0 +1,196 @@
+	)
+	for rows.Next() {
+		err = rows.Scan(&variable, &maxConn)
+		if err != nil {
+			markCheckError(result, err)
+			return result
+		}
+	}
+	rows.Close()
+	if maxConn == 0 {
+		c.unlimitedConn = true
+		result.State = StateSuccess
+		return result
+	}
+	processRows, err = baseConn.QuerySQL(tcontext.NewContext(ctx, log.L()), "SHOW PROCESSLIST")
+	if err != nil {
+		markCheckError(result, err)
+		return result
+	}
+	defer processRows.Close()
+	usedConn := 0
+	for processRows.Next() {
+		usedConn++
+	}
+	usedConn -= 1 // exclude the connection used for show processlist
+	log.L().Debug("connection checker", zap.Int("maxConnections", maxConn), zap.Int("usedConnections", usedConn))
+	neededConn := c.getConfigConn(c.stCfgs)
+	if neededConn > maxConn {
+		// nonzero max_connections and needed connections exceed max_connections
+		// FYI: https://github.com/pingcap/tidb/pull/35453
+		// currently, TiDB's max_connections is set to 0 representing unlimited connections,
+		// while for MySQL, 0 is not a legal value (never retrieve from it).
+		result.Errors = append(
+			result.Errors,
+			NewError(
+				"checked database's max_connections: %d is less than the number %s needs: %d",
+				maxConn,
+				c.workerName,
+				neededConn,
+			),
+		)
+		result.Instruction = "set larger max_connections or adjust the configuration of dm"
+		result.State = StateFailure
+	} else {
+		// nonzero max_connections and needed connections are less than or equal to max_connections
+		result.State = StateSuccess
+		if maxConn-usedConn < neededConn {
+			result.State = StateWarning
+			result.Instruction = "set larger max_connections or adjust the configuration of dm"
+			result.Errors = append(
+				result.Errors,
+				NewError(
+					"database's max_connections: %d, used_connections: %d, available_connections: %d is less than %s needs: %d",
+					maxConn,
+					usedConn,
+					maxConn-usedConn,
+					c.workerName,
+					neededConn,
+				),
+			)
+		}
+	}
+	return result
+}
+
+type LoaderConnNumberChecker struct {
+	connNumberChecker
+}
+
+func NewLoaderConnNumberChecker(targetDB *conn.BaseDB, stCfgs []*config.SubTaskConfig) RealChecker {
+	return &LoaderConnNumberChecker{
+		connNumberChecker: newConnNumberChecker(targetDB, stCfgs, func(stCfgs []*config.SubTaskConfig) int {
+			loaderConn := 0
+			for _, stCfg := range stCfgs {
+				// loader's worker and checkpoint (always keeps one db connection)
+				loaderConn += stCfg.LoaderConfig.PoolSize + 1
+			}
+			return loaderConn
+		}, "loader"),
+	}
+}
+
+func (l *LoaderConnNumberChecker) Name() string {
+	return "loader_conn_number_checker"
+}
+
+func (l *LoaderConnNumberChecker) Check(ctx context.Context) *Result {
+	result := l.check(ctx, l.Name())
+	if !l.unlimitedConn && result.State == StateFailure {
+		// if the max_connections is set as a specific number
+		// and we failed because of the number connecions needed is smaller than max_connections
+		for _, stCfg := range l.stCfgs {
+			if stCfg.NeedUseLightning() {
```

Comment:
not-use-lightning should have higher priority than use-lightning

Reply:
Would you mind suggesting more about this?

Reply:
sorry I forget all load units are connected to one downstream, so if any one of them is lightning it will become a "soft" limit.